src/tree_schema.h

```diff
@@ -1568,69 +1606,166 @@ struct lysc_node {
     void *priv;                      /**< private arbitrary user data, not used by libyang */
 };
 
+struct lysc_node_action_inout {
+    union {
+        struct lysc_node node;       /**< implicit cast for the members compatible with ::lysc_node */
+        struct {
+            uint16_t nodetype;       /**< LYS_INPUT or LYS_OUTPUT */
+            uint16_t flags;          /**< [schema node flags](@ref snodeflags) */
+            uint8_t hash[LYS_NODE_HASH_COUNT]; /**< schema hash required for LYB printer/parser */
+            struct lys_module *module; /**< module structure */
+            struct lysc_node *parent; /**< parent node (NULL in case of top level node) */
+            struct lysc_node *next;  /**< next sibling node (NULL if there is no one) */
+            struct lysc_node *prev;  /**< pointer to the previous sibling node \note Note that this pointer is
+                                          never NULL. If there is no sibling node, pointer points to the node
+                                          itself. In case of the first node, this pointer points to the last
+                                          node in the list. */
+            const char *name;        /**< Notification name (mandatory) */
+            const char *dsc;         /**< description */
+            const char *ref;         /**< reference */
+            struct lysc_ext_instance *exts; /**< list of the extension instances ([sized array](@ref sizedarrays)) */
+            struct lysc_when **when; /**< list of pointers to when statements ([sized array](@ref sizedarrays)) */
+            void *priv;              /** private arbitrary user data, not used by libyang */
+        };
+    };
+
+    struct lysc_node *child;          /**< first child node (linked list) */
+    struct lysc_must *musts;         /**< list of must restrictions ([sized array](@ref sizedarrays)) */
+};
```

Comment:
The `lysc_node` and `lysp_node` structures are part of the API, so there shouldn't be any changes (each change here means moving to a new SONAME and losing backward compatibility).
Having the common members explicitly even in the derived structures allows providing more precise documentation regarding the content of each member in a specific node type (expected flags, not used members, etc.). The `name` in the augment statement has actually the meaning of `nodeid` (target), which can be explained this way.
I believe that this mess in the definition is worthwhile.

Reply:
Has libyang2 actually been released yet though? That's what I meant about now being the time to change. To your second point about documenting different meanings, which I'm more sympathetic to, if they have different meanings though, they maybe shouldn't be in the common node structure.

Reply:
No, it wasn't released yet, that's why we are able to do this kind of change.

Regarding the different meanings of the members - it is not a common situation, but there are few such cases. The reason why the structures are supposed to be somehow compatible is that they all can appear in schema tree (parsed or compiled), so they are hidden behind the `parent` or `child` pointers. But you are right, that now after unifying all the relevant structures, we should again look at the members if they are really necessary to be in the common structure.

I still believe that better documentation and easier use of the members is sufficient justification for this solution. It seems that there are benefits (more freedom regarding accessing the members, better understandability thanks to docs) for users and the only problem is a little bit more challenging maintenance in case of structure change for us (libyang developers). The first case is supposed to be common with a larger group of affected people, while the second case is rare with only a few affected people.

Reply:
I think you will find that most people are confused by the union. It's sort of abusing the language to make it 2 keystrokes less to type. The documentation for the common fields goes on the common structure and that's where I would look for it. Again, I think you have a workable solution, but I think KISS and "do the obvious thing" would not use a union.